### ISSUE TITLE
No write to non-contiguous views

### ISSUE THREAD

**Issue by michel2323** [JuliaGPU/AMDGPU.jl#140]
The following MWE shows that non-contiguous views do not seems to work as intended. They seem to be read-only. However, contiguous views work fine. 

```julia
julia> idx = ROCVector{Int}([1,3])
3-element ROCVector{Int64}:
 1
 3
julia> buf = ROCVector{Float64}([1.0,1.0,1.0])
3-element ROCVector{Float64}:
 1.0
 1.0
 1.0
julia> vbuf = view(buf,idx)
2-element view(::ROCVector{Float64}, [1, 3]) with eltype Float64:
 1.0
 1.0
julia> vbuf .= 2.0 # <--- WON'T WORK
2-element view(::ROCVector{Float64}, [1, 3]) with eltype Float64:
 1.0
 1.0
julia> buf .= 2.0
3-element ROCVector{Float64}:
 2.0
 2.0
 2.0
julia> vbuf # <--- UPDATING PARENT WORKED
2-element view(::ROCVector{Float64}, [1, 3]) with eltype Float64:
 2.0
 2.0
julia> cvbuf = view(buf,1:2)
2-element ROCVector{Float64}:
 2.0
 2.0
julia> cvbuf .= 3 # <--- CONTIGUOUS VIEW WORKS
2-element ROCVector{Float64}:
 3.0
 3.0
julia> buf
3-element ROCVector{Float64}:
 3.0
 3.0
 2.0
```

**Comment by vchuravy** [comment]
So:

```
using AMDGPU
using LinearAlgebra

C = ROCArray(zeros(2, 2))

function kernel(data)
  data[1] = 1.0
  nothing
end

wait(@roc groupsize=1 kernel(view(C, diagind(C))))
```

already fails.

**Comment by vchuravy** [comment]
Further reduced to 

```
using AMDGPU

C = ROCArray(zeros(4))
vC = view(C, 1:3:4)

@noinline escape(x) = nothing
function kernel(data)
  escape(length(data))
  nothing
end

wait(@roc groupsize=1 kernel(vC))
```

**Comment by vchuravy** [comment]
Further reduced to:

```
using AMDGPU


function kernel(data)
  I = first(data.indices)
  @rocprintf "start %d, stop %d, step %d\n" I.start I.stop I.step
  nothing
end

C = ROCArray(zeros(4))
vC = view(C, 1:3:4)

wait(@roc groupsize=1 kernel(vC))
```

yielding:

```
[vchuravy@node2002 amd]$ julia-1.6.3/bin/julia test.jl 
^[[O┌ Warning: oclc_wavefrontsize64_on.bc not found
└ @ AMDGPU /nobackup/users/vchuravy/src/AMDGPU/src/device/runtime.jl:154
'+fp64-fp16-denormals' is not a recognized feature for this target (ignoring feature)
'-fp32-denormals' is not a recognized feature for this target (ignoring feature)
'+fp64-fp16-denormals' is not a recognized feature for this target (ignoring feature)
'-fp32-denormals' is not a recognized feature for this target (ignoring feature)
'+fp64-fp16-denormals' is not a recognized feature for this target (ignoring feature)
'-fp32-denormals' is not a recognized feature for this target (ignoring feature)
'+fp64-fp16-denormals' is not a recognized feature for this target (ignoring feature)
'-fp32-denormals' is not a recognized feature for this target (ignoring feature)
start 0, stop 0, step 0
```

**Comment by vchuravy** [comment]
I should note that passing a `StepRange` in and printing it is correct. So the issue seems to be related to depth/complexity of the struct.

**Comment by vchuravy** [comment]
Fixed in #171
